### DISCUSSION TITLE
Is there a way to use a transparent texture in Beer/Malt?

### DISCUSSION THREAD

**Discussion by MeganDrawsaLot** [bnpr/Malt#51] [Q&A] [answered]
I've been trying to see if I can use transparent textures while using the Malt render engine.  I've tried using the Basic Texture Shader, but there is no transparency when I use a .png.  Is there a way I could change the shaders code to achieve this, and make a texture visible without having to add a light to the scene?
I feel like this could be very helpful for adding painted layers, like for highlights in hair, or tattoo designs.  Right now I'm trying to have my logo next to the character I'm working on in blender.

For the example I pushed the logo to sit in front of the character more, to try and be clearer about what I mean.

![Illustration3](https://user-images.githubusercontent.com/77802440/105613569-4b4e5d00-5d91-11eb-9fbd-6a7d8d2b0b44.png)

**Comment by pragma37** [comment, marked as answer]
>  I've tried using the Basic Texture Shader, but there is no transparency when I use a .png.

If you look at the code, the line 13 is reading only the rgb channels:
```glsl
vec3 diffuse_color = texture(diffuse_texture, UV[uv_index]).rgb;
```
If you just want a flat texture without lighting, you could use a shader that simply outputs the full texture rgba values as the pixel color output:
```glsl
#include "Pipelines/NPR_Pipeline.glsl"

uniform sampler2D diffuse_texture;
uniform int uv_index = 0;

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    PO.color = texture(diffuse_texture, S.uv[uv_index]);
}
```

If you want to adapt the basic_texture.mesh.glsl you could just use the texture alpha as pixel color output alpha:
```glsl
#include "Pipelines/NPR_Pipeline.glsl"

uniform vec3 ambient_color = vec3(0.1,0.1,0.1);
uniform sampler2D diffuse_texture;
uniform int uv_index = 0;
uniform vec3 specular_color = vec3(1.0,1.0,1.0);
uniform float roughness = 0.5;

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    vec4 diffuse_color = texture(diffuse_texture, S.uv[uv_index]);
    vec3 diffuse = diffuse_color.rgb * get_diffuse();
    vec3 specular = specular_color * get_specular(roughness);
    vec3 color = ambient_color + diffuse + specular;

    PO.color.rgb = color;
    PO.color.a = diffuse_color.a;
}
```
By default, only pixels with alpha 0 will be transparent and the rest will be fully opaque. If you want alpha blending enable the Transparency options in the material UI.

**Reply by MeganDrawsaLot** [reply]
Okay, I got it to work!
I'm also trying to figure out how to use a texture and line shader at the same time.  Maybe I'm not putting the code in the right order? 
 'Always will appreciate any pointers that can be given.
Thank you for the help that you've provide so far! ヽ(º ▽ º )ノ✨

**Reply by pragma37** [reply]
> I'm also trying to figure out how to use a texture and line shader at the same time. Maybe I'm not putting the code in the right order?

Do you get any errors?
If you post what you have so far I can point you in the right direction.

**Reply by MeganDrawsaLot** [reply]
This is what I have set up so far, and yes there is a wall of error text in blender.  It's kind of intimidating. (^-^;)>

```
//Copyright (c) 2020 BlenderNPR and contributors. MIT license.

#include "Pipelines/NPR_Pipeline.glsl"

uniform vec3 ambient_color = vec3(0.1,0.1,0.1);
uniform vec3 diffuse_color = vec3(1.0,0.1,0.1);
uniform vec3 specular_color = vec3(1.0,1.0,1.0);
uniform float roughness = 0.5;

uniform bool line_world_space = false;

uniform vec4 line_color = vec4(0.0,0.0,0.0,1.0);

uniform float line_id_boundary_width = 1.;

uniform float line_depth_threshold_min = 0.5;
uniform float line_depth_threshold_max = 2.0;
uniform float line_depth_width_min = 0.5;
uniform float line_depth_width_max = 2.0;

uniform float line_angle_threshold_min = 0.5;
uniform float line_angle_threshold_max = 1.5;
uniform float line_angle_width_min = 0.5;
uniform float line_angle_width_max = 2.0;


void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    vec3 diffuse = diffuse_color * get_diffuse_half();
    vec3 specular = specular_color * get_specular(roughness);
    vec3 color = ambient_color + diffuse + specular;

    float line = get_line_advanced(
        line_id_boundary_width,
        line_depth_threshold_min, line_depth_threshold_max, line_depth_width_min, line_depth_width_max,
        line_angle_threshold_min, line_angle_threshold_max, line_angle_width_min, line_angle_width_max);
    
    if(line_world_space)
    {
        line /= pixel_world_size_at(pixel_depth());
        // Here we divide by 100 just to make line size more consistent between modes,
        // but this is a completely optional hand-picked number
        line /= 100;
    }

    PO.color.rgb = color;
    PO.line_color = line_color;
    PO.line_width = line * S.color[0].r;
    
}
#include "Pipelines/NPR_Pipeline.glsl"

uniform vec3 ambient_color = vec3(0.1,0.1,0.1);
uniform sampler2D diffuse_texture;
uniform int uv_index = 0;
uniform vec3 specular_color = vec3(1.0,1.0,1.0);
uniform float roughness = 0.5;

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    vec4 diffuse_color = texture(diffuse_texture, S.uv[uv_index]);
    vec3 diffuse = diffuse_color.rgb * get_diffuse();
    vec3 specular = specular_color * get_specular(roughness);
    vec3 color = ambient_color + diffuse + specular;

    PO.color.rgb = color;
    PO.color.a = diffuse_color.a;
```

**Reply by pragma37** [reply]
Oh! I see, yeah you can't just put one shader after another. You have to make one shader that implements both functionalities.

So, to begin with, you can have only one COMMON_PIXEL_SHADER function. 
The same goes for uniforms,  you can't declare the same uniform twice. Notice how you have the ambient_color, specular_color and roughness uniforms duplicated.

What you have at the moment is basically this:

```
uniforms for the surface (from the line shader example)
uniforms for the line (from the line shader example)

Common Pixel Shader function
{
    Surface color calculation (from the line shader example)
    
    Line calculation (from the line shader example)
    
    Set the color output (from the line shader example)
    
    Set the line output (from the line shader example)
}

uniforms for the surface (from the texture shader example)

Common Pixel Shader function
{
    Surface color calculation (from the texture shader example)
    
    Set the color output (from the texture shader example)
}
```

But what you actually want is this:

```
uniforms for the surface (from the texture shader example)
uniforms for the line (from the line shader example)

Common Pixel Shader function
{
    Surface color calculation (from the texture shader example)
    
    Line calculation (from the line shader example)
    
    Set the color output (from the texture shader example)
    
    Set the line output (from the line shader example)
}
```
Does that makes sense?

Whenever you see an error on the material panel,  just start with the first one and go from there.
They tell you at what line it is and what the error is.
If the error description doesn't makes sense, 90% of the time is just that you forgot to put a semicolon or to close a parenthesis or a brace.

If you can't get it to work I can just post the working shader, but try to do it yourself first. 👍

**Reply by MeganDrawsaLot** [reply]
I think I am getting the general idea of it, my brain just needs to learn how to take in and sort all the new information.
Thank you once again for the help!  I think I'll be able to figure it out, but will definitely ask more questions if I get majorly stuck. :)

**Reply by MeganDrawsaLot** [reply]
I got it!!! Yay!

Is there a glsl code I can use to get the grease-pencil to show up behind the transparent texture as well.
...I've uh, got another question as well but I don't want to take up to much of your time. ^^;


Here's the code so far. 

```

//Copyright (c) 2020 BlenderNPR and contributors. MIT license.

#include "Pipelines/NPR_Pipeline.glsl"

uniform vec3 ambient_color = vec3(0.1,0.1,0.1);
uniform vec3 diffuse_color = vec3(1.0,0.1,0.1);

uniform sampler2D diffuse_texture;
uniform int uv_index = 0;

uniform vec3 specular_color = vec3(1.0,1.0,1.0);
uniform float roughness = 0.5;

uniform bool line_world_space = false;

uniform vec4 line_color = vec4(0.0,0.0,0.0,1.0);

uniform float line_id_boundary_width = 1.;

uniform float line_depth_threshold_min = 0.5;
uniform float line_depth_threshold_max = 2.0;
uniform float line_depth_width_min = 0.5;
uniform float line_depth_width_max = 2.0;

uniform float line_angle_threshold_min = 0.5;
uniform float line_angle_threshold_max = 1.5;
uniform float line_angle_width_min = 0.5;
uniform float line_angle_width_max = 2.0;


void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{

    vec4 diffuse_color = texture(diffuse_texture, S.uv[uv_index]);
    vec3 diffuse = diffuse_color.rgb * get_diffuse();
    vec3 specular = specular_color * get_specular(roughness);
    vec3 color = ambient_color + diffuse + specular;

    
    
    float line = get_line_advanced(
        line_id_boundary_width,
        line_depth_threshold_min, line_depth_threshold_max, line_depth_width_min, line_depth_width_max,
        line_angle_threshold_min, line_angle_threshold_max, line_angle_width_min, line_angle_width_max);
    
    if(line_world_space)
    {
        line /= pixel_world_size_at(pixel_depth());
        // Here we divide by 100 just to make line size more consistent between modes,
        // but this is a completely optional hand-picked number
        line /= 100;
    }

    PO.color = texture(diffuse_texture, S.uv[uv_index]);
    PO.line_color = line_color;
    PO.line_width = line * S.color[0].r;
}



```

**Reply by MeganDrawsaLot** [reply]
Here's a picture of what I was talking about with the grease pencil.

![image](https://user-images.githubusercontent.com/77802440/105773057-2db7f980-5f31-11eb-8f4a-ff851bfe5194.png)

**Reply by pragma37** [reply]
Yay! Congrats!

There's no way to fix that grease pencil issue afaik, sorry. I think even EEVEE and Cycles don't support that at the moment.
Basically grease pencil does its own rendering, so it's a take it or leave it situation.

I think you can convert grease pencil objects to mesh objects, though.

**Reply by MeganDrawsaLot** [reply]
Converting it to mesh was my next thought! :)

Okay here's another idea; is there any way to make a transparent cube, that has line art, in Malt?  This is more for curiosities sake.  I Made a cube and put my logo on it and found out that shader for the lines work around the alpha channel/will outline the logo itself.  I was wondering if the cube could be outlined while still retaining the transparent mesh.  I don't know what I'd use this method for (yet) but my brain is kind of itching to know if it can be done.

Thanks again for your prompt answers!
